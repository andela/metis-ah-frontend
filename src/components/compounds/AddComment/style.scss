
#add-comment-view {
  border: 1px solid rgba(0,0,0,0.09);
  border-radius: 4px;
  background: white;
  margin-bottom: 50px;
  padding-bottom: 10px;

  .input-container {
    text-align: right;
    margin-top: 3px;
    margin-right: 10px;
    margin-left: 10px;
  }

  textarea {
    width: 100%;
    height: 150px;;
    padding: 20px 5px 5px;
    font-size: 17px;
    outline: 0;
    background: white;
    border: 0;
    margin: 0;
    border-top-right-radius: 4px;
    border-top-left-radius: 4px;
    border-bottom: 1px solid rgba(0,0,0,0.09);

    &::placeholder {
      color: rgba(0,0,0,0.4);
    }
  }

  input[type="submit"] {
    border: 0;
    background: #4dd6a5;
    color: white;
    font-weight: 600;
    margin-top: 0px;
    padding: 8px 25px;
    border-radius: 20px;
    cursor: pointer;
    font-size: 16px;
  }
  input[type="submit"]:hover {
    background: #11b179;
  }
}