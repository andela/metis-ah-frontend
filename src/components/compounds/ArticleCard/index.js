import React from 'react';
import PropTypes from 'prop-types';
import { dayHelper, monthHelper } from '../../util/date';

const ArticleCard = ({
  article, image, firstname, lastname,
}) => {
  const {
    title, createdAt, description, imageUrl, ratings
  } = article;
  return (
    <div className="column is-one-third">
      <div className="ArticleCard">
        <div
          className="ArticleCard__banner"
          style={{
            backgroundImage: `linear-gradient(to bottom, transparent, #000000),url(${imageUrl})`
          }}
        >
          <div className="ArticleCard__BannerContent">
            <h1 className="ArticleCard__BannerTitle">
              {title}
            </h1>
            <p className="ArticleCard__date">
              {monthHelper(createdAt)}
              {' '}
              {dayHelper(createdAt)}
            </p>
          </div>
          <div className="ArticleCard__overlay is-overlay">
            <div className="ArticleCard__OverlayContent">
              <h1 className="ArticleCard__OverlayTitle">{title}</h1>
              <p className="ArticleCard__OverlayText">
                {description}
              </p>
              <p className="ArticleCard__OverlayDate">
                {monthHelper(createdAt)}
                {' '}
                {dayHelper(createdAt)}
              </p>
            </div>
          </div>
        </div>
        <div className="ArticleCard__footer">
          <div className="columns is-mobile">
            <div className="column is-one-fifth PaddingRight--0 Margin--auto">
              <div className="ArticleCard__FooterThumbnail">
                <img src={image} alt="Article author" />
              </div>
            </div>
            <div className="column is-paddingless Margin--auto is-three-fifths">
              <div className="ArticleCard__FooterName">
                {firstname}
                {' '}
                {lastname}
              </div>
            </div>
            <div className="column Margin--auto is-paddingless">
              <div className="ArticleCard__FooterLikes">
                <i className="fas fa-thumbs-up fa-2x" />
                <span>{ratings || 0}</span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>);
};

ArticleCard.propTypes = {
  article: PropTypes.objectOf(PropTypes.any).isRequired,
  firstname: PropTypes.string.isRequired,
  lastname: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
};

export default ArticleCard;
