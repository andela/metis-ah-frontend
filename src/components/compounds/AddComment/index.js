import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';

import { addComment } from 'Actions/commentActions';

import InputField from 'Atoms/InputField/InputField';
import { showModal } from 'Actions/authUser';
import Alert from 'Utils/helpers/alert';
import './style.scss';

export class AddComment extends React.Component {
  state = {
    content: ''
  };


  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value
    });
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const { content } = this.state;
    const { isAuth, showModal } = this.props;

    if (content.length < 2) {
      return Alert.error('Please, enter a comment');
    }
    if (!isAuth) {
      Alert.info('Hi there, you need to be registered or logged in to comment');
      return showModal();
    }
    const { match, addComment } = this.props;
    const { articleId } = match.params;
    return addComment(content, articleId).then(() => {
      this.setState({
        content: ''
      });
    });
  }

  render() {
    const { content } = this.state;
    const { isCreateCommentLoading } = this.props;
    return (
      <div id="add-comment-view">
        <form onSubmit={this.handleSubmit}>
          <textarea
            onChange={this.handleChange}
            name="content"
            placeholder="Write a comment"
            rows="10"
            cols="30"
            value={content}
          />
          <div className="input-container">
            <i id="spinner" className={isCreateCommentLoading ? 'fas fa-spinner fa-spin' : ''} />
            <InputField
              type="submit"
              value="Comment"
              style={isCreateCommentLoading ? { opacity: 0.3 } : null}
              disabled={isCreateCommentLoading}
            />
          </div>
        </form>
      </div>
    );
  }
}

AddComment.defaultProps = {
  isCreateCommentLoading: false
};

AddComment.propTypes = {
  match: PropTypes.objectOf(PropTypes.any).isRequired,
  addComment: PropTypes.func.isRequired,
  showModal: PropTypes.func.isRequired,
  isAuth: PropTypes.bool.isRequired,
  isCreateCommentLoading: PropTypes.bool
};

const mapStateToProps = state => ({
  isAuth: state.authUser.isAuthenticated,
  isCreateCommentLoading: state.commentReducer.isCreateCommentLoading
});

export default withRouter(connect(
  mapStateToProps,
  { addComment, showModal }
)(AddComment));
