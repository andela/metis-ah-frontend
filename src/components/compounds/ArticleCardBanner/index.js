import React, { Component } from 'react';
import PropTypes from 'prop-types';

import './style.scss';

const style = image => ({
  background: `linear-gradient(to bottom, transparent, transparent, #000000), url('${image}')`
});

const descriptionStyle = () => ({
  display: 'block',
});

class ArticleCardBanner extends Component {
  state = {
    hover: false,
  }

  clearStyle = () => {
    this.setState({
      hover: true
    });
  };

  returnStyle = () => {
    this.setState({
      hover: false
    });
  };

  render() {
    const { hover } = this.state;
    const { item } = this.props;
    return (
      <div
        className="ArticleCard__banner article-card-banner"
        style={hover ? {} : style(item.banner)}
        onMouseEnter={this.clearStyle}
        onMouseLeave={this.returnStyle}
      >
        <div className="ArticleCard__BannerContent">
          <h3 className="ArticleCard__BannerTitle">
            {item.title}
          </h3>
          <p
            className="ArticleCard__BannerDescription"
            style={!hover ? {} : descriptionStyle()}
          >
            {item.description}
          </p>
          <p className="ArticleCard__date">
            {item.date}
          </p>
        </div>
      </div>
    );
  }
}

ArticleCardBanner.propTypes = {
  item: PropTypes.shape({
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
    likes: PropTypes.number.isRequired,
    banner: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    authorImg: PropTypes.string.isRequired,
  }).isRequired,
};

export default ArticleCardBanner;
