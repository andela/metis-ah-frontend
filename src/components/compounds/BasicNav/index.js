import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';

import './style.scss';

const BasicNav = ({
  logo, brandName, classes = 'navbar', id = 'navbar-plain', ...props
}) => (
  <Fragment>
    <nav id={id} {...props} className={classes}>
      <div className="container">
        <Link className="navbar-item" to="/">
          <img id="logo" src={logo} alt="Authors haven logo" />
          <h1 className="brand">{brandName}</h1>
        </Link>
      </div>
    </nav>
  </Fragment>
);

export default BasicNav;
