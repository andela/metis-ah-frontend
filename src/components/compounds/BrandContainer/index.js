import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import logo from 'Images/logo.png';

import './style.scss';

const BrandContainer = (props) => {
  const { showMenu, menu } = props;
  return (
    <div className="navbar-brand brand header-brand-container">
      <Link className="navbar-item" to="/">
        <img id="logo" src={logo} alt="Authors haven logo" />
        <h1 className="title">Author&apos;s Haven</h1>
      </Link>
      <button
        type="button"
        className={`navbar-burger burger' ${menu ? 'is-active' : ''}' bug`}
        aria-label="menu"
        aria-expanded="false"
        data-target="profile-nav"
        onClick={showMenu}
      >
        <span aria-hidden="true" />
        <span aria-hidden="true" />
        <span aria-hidden="true" />
      </button>
    </div>
  );
};

BrandContainer.propTypes = {
  showMenu: PropTypes.func.isRequired,
  menu: PropTypes.bool.isRequired,
};

export default BrandContainer;
