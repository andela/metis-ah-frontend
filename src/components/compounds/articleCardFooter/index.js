import React from 'react';
import PropTypes from 'prop-types';

import './style';

const computeLikes = (likes) => {
  if (likes >= 1000) {
    let num = likes / 1000;
    num = num.toString();
    num += 'K';

    return num;
  }
};

const articleCardFooter = (props) => {
  const { item } = props;
  return (
    <div className="ArticleCard__footer">
      <div className="ArticleCard__FooterThumbnail">
        <img src={item.authorImg} alt="Article author" />
      </div>
      <div className="ArticleCard__FooterName">
        {item.author}
      </div>
      <div className="ArticleCard__FooterLikes">
        <i className="fas fa-thumbs-up fa-2x" />
        <span>{computeLikes(item.likes)}</span>
      </div>
    </div>
  );
};

articleCardFooter.propTypes = {
  item: PropTypes.shape({
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
    likes: PropTypes.number.isRequired,
    banner: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    authorImg: PropTypes.string.isRequired,
  }).isRequired,
};

export default articleCardFooter;
