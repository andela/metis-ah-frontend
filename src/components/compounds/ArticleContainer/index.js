
import React, { Component } from 'react';
import { withRouter, Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ArticleCard from 'Components/atoms/ArticleCard';
import { getArticle } from 'Actions/article';
import sortArrayContent from '../../util/helper';
import './style.scss';

class ArticleContainer extends Component {
  constructor(props) {
    super(props);
    const {
      match: { params: category }
    } = this.props;
    this.state = {
      category
    };
  }

  componentDidMount() {
    const { getArticles, match, categories } = this.props;
    const heroContent = sortArrayContent(categories, match.params.category);
    getArticles(match.params.category, heroContent);
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    const {
      match: {
        params: { category }
      }
    } = nextProps;
    const { categories } = nextProps;
    const heroContent = sortArrayContent(categories, category);
    if (prevState.category.category !== nextProps.match.params.category) {
      nextProps.getArticles(category, heroContent);
      return {
        category: { category }
      };
    }
    return null;
  }

  render() {
    const { articles, loading } = this.props;

    const mappedArticleCards = articles.map((article) => {
      const {
        id,
        imageUrl,
        slug,
        title,
        description,
        createdAt,
        User: { firstname, lastname }
      } = article;
      return (
        <ArticleCard
          key={slug}
          id={id}
          img={imageUrl}
          title={title}
          description={description}
          name={`${firstname} ${lastname}`}
          time={createdAt}
        />
      );
    });
    return (
      <div className="container article-inner-container">
        {mappedArticleCards}
        {articles.length < 1 && (loading && <div className="article-loader" />)}
        {articles.length < 1
          && (!loading && (
            <div className="article-container-card">
              <i className="fas fa-exclamation notification-icon" />
              <p>
                Oops!, No Article currently exits in this category.
                <br />
                We would like you to be the first to write in this category
              </p>
              <br />
              <Link to="/articles/new">WRITE</Link>
            </div>
          ))}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  loading: state.article.loading,
  articles: state.article.articles,
  categories: state.categories.categories
});

const mapDispatchToProps = dispatch => ({
  getArticles: (categoryId, heroContent) => dispatch(getArticle(categoryId, heroContent))
});

ArticleContainer.propTypes = {
  articles: PropTypes.arrayOf(PropTypes.object),
  loading: PropTypes.bool.isRequired,
  match: PropTypes.objectOf(PropTypes.any).isRequired,
  getArticles: PropTypes.func.isRequired,
  categories: PropTypes.arrayOf(PropTypes.object).isRequired
};

ArticleContainer.defaultProps = {
  articles: []
};
export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(ArticleContainer)
);
