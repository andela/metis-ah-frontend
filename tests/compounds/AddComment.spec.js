import React from 'react';
import { shallow } from 'enzyme';


import { AddComment } from 'Components/compounds/AddComment';


const mockFn = jest.fn();


const props = {
  match: {
    params: {
      token: 'k4tkwertlk554kjlkh4235lkjl6ghjgf56fhf352'
    }
  },
  showModal: mockFn,
  addComment: jest.fn(() => Promise.resolve()),
  isAuth: false,
  isCreateCommentLoading: false
};


describe('ADD COMMENT TESTS', () => {
  test('renders without crashing', () => {
    const wrapper = shallow(<AddComment {...props} />);
    expect(wrapper.exists()).toBe(true);
  });

  test('should match snapshot', () => {
    const wrapper = shallow(<AddComment {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should call showModal form submits and user is not logged in', () => {
    const wrapper = shallow(<AddComment {...props} />);

    const event = {
      preventDefault: mockFn
    };

    wrapper.setState({
      content: 'Nice work'
    });

    wrapper.find('form').simulate('submit', event);
    expect(wrapper.instance().props.showModal).toHaveBeenCalled();
  });

  test('should call handleSubmit method on form submit', () => {
    const wrapper = shallow(<AddComment {...props} />);

    const event = {
      preventDefault: mockFn
    };

    wrapper.setState({
      content: 'Nice work'
    });

    wrapper.setProps({
      isAuth: true
    });

    wrapper.find('form').simulate('submit', event);
    expect(wrapper.instance().props.addComment).toHaveBeenCalled();
  });

  test('should set comment content on input change', () => {
    const wrapper = shallow(<AddComment {...props} />);

    const event = {
      target: {
        name: 'content',
        value: 'I love your article, keep up the good job'
      }
    };
    const commentTextArea = wrapper.find('[name="content"]');
    commentTextArea.simulate('change', event);

    expect(wrapper.instance().state.content).toEqual('I love your article, keep up the good job');
  });

  test('should should freeze when loading', () => {
    const wrapper = shallow(<AddComment {...props} />);

    wrapper.setProps({ isCreateCommentLoading: true });
    expect(wrapper.instance().props.isCreateCommentLoading).toBe(true);
  });

  test('should match snapshot when content is empty', () => {
    const wrapper = shallow(<AddComment {...props} />);

    const event = {
      preventDefault: mockFn
    };
    wrapper.setState({ content: '' });

    wrapper.find('form').simulate('submit', event);
    expect(wrapper).toMatchSnapshot();
  });
});
