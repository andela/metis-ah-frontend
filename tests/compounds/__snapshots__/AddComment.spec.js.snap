// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ADD COMMENT TESTS should match snapshot 1`] = `
<div
  id="add-comment-view"
>
  <form
    onSubmit={[Function]}
  >
    <textarea
      cols="30"
      name="content"
      onChange={[Function]}
      placeholder="Write a comment"
      rows="10"
      value=""
    />
    <div
      className="input-container"
    >
      <i
        className=""
        id="spinner"
      />
      <InputField
        disabled={false}
        label=""
        name=""
        style={null}
        type="submit"
        value="Comment"
      />
    </div>
  </form>
</div>
`;

exports[`ADD COMMENT TESTS should match snapshot when content is empty 1`] = `
<div
  id="add-comment-view"
>
  <form
    onSubmit={[Function]}
  >
    <textarea
      cols="30"
      name="content"
      onChange={[Function]}
      placeholder="Write a comment"
      rows="10"
      value=""
    />
    <div
      className="input-container"
    >
      <i
        className=""
        id="spinner"
      />
      <InputField
        disabled={false}
        label=""
        name=""
        style={null}
        type="submit"
        value="Comment"
      />
    </div>
  </form>
</div>
`;
